src/requestor.ts

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {AppAuthError} from './errors';
+import {StringMap} from './types';
+
+export interface RequestorBaseSettings {
+  /** URL to request */
+  url: string;
+  /** HTTP method, as uppercase string */
+  method: string;
+  /** Expected response data type. Currently, requestors only need to handle JSON responses. */
+  dataType: 'json';
+  /** Any custom headers or header overrides on the request */
+  headers?: StringMap;
+}
+
+export interface RequestorGetSettings extends RequestorBaseSettings { method: 'GET'; }
+
+export interface RequestorPostSettings extends RequestorBaseSettings {
+  method: 'POST';
+  /**
+   * data to send on request as URL-encoded form data. If an object, will be treated as
+   * key=value form data.
+   */
+  data: StringMap;
```

Comment:
So `GET` requests could have a payload as well. There is nothing in the spec that says that configuration fetches may not require query params. So we should support that.

Also `data` does not need to be constrained to a `StringMap`. It could be a `{}` type. That will avoid the force cast to `any` that you are doing.

I think choosing an implicit `dataType` = `json` in `RequestorBaseSettings` might not be great for configuration fetches which might only support `application/x-www-form-urlencoded`.

Reply:
`dataType` defines the response content type, not the request content type. All current usage returns JSON, hence me restricting it in the interface.

`GET` requests cannot have a body per HTTP spec (or rather, they are defined to have no documented behavior and to cause many implementations to reject the request). The two interfaces make `GET` disallow request data, and `POST` require it, rather than making it optional for both.

The URL could have query parameters of course but `data` is being used to generate a form-encoded request body, not to modify the URL.

`data` was only used as either a StringMap or a string - I simply changed the one usage as a string to use StringMap instead.